src/Symfony/Component/HttpFoundation/Request.php

```diff
@@ -1649,32 +1662,40 @@ public function getLanguages(): array
         $this->languages = [];
         foreach ($languages as $acceptHeaderItem) {
             $lang = $acceptHeaderItem->getValue();
-            if (str_contains($lang, '-')) {
-                $codes = explode('-', $lang);
-                if ('i' === $codes[0]) {
-                    // Language not listed in ISO 639 that are not variants
-                    // of any listed language, which can be registered with the
-                    // i-prefix, such as i-cherokee
-                    if (\count($codes) > 1) {
-                        $lang = $codes[1];
-                    }
-                } else {
-                    for ($i = 0, $max = \count($codes); $i < $max; ++$i) {
-                        if (0 === $i) {
-                            $lang = strtolower($codes[0]);
-                        } else {
-                            $lang .= '_'.strtoupper($codes[$i]);
-                        }
-                    }
-                }
-            }
+            $lang = $this->formatLocale($lang);
 
             $this->languages[] = $lang;
         }
 
         return $this->languages;
     }
 
+    private static function formatLocale(string $locale): string
+    {
+        if (!str_contains($locale, '-')) {
+            return $locale;
+        }
+        $codes = explode('-', $locale);
+        if ('i' === $codes[0]) {
+            // Language not listed in ISO 639 that are not variants
+            // of any listed language, which can be registered with the
+            // i-prefix, such as i-cherokee
+            if (\count($codes) > 1) {
+                $locale = $codes[1];
```

Comment:
```suggestion
                return $codes[1];
```

Reply:
Hi @stloyd,

Many thanks for the suggestion.
I will not commit it as I am working on this part that needs other modifications.
I noticed issues when variant are present in the locale e.g. zh-hans-tw (simplified Chinese from Taiwan) or when components/extensions follow the region (`pt-BR-u-ca-gregory-nu-latn` for Brazilian Portuguese with use of the Gregorian calendar and Latin numbering).
This function should return `zh_Hans_TW` or `pt_BR`, but it does not and an exact match may be missed because of this.
For `zh_Hans_TW`, it could fallback to `zh_TW` and `zh` but this is not the case at the moment.

@welcoMattic I think we could add a class to handle these locale parts in the Translation component. This will help the Request object handle the languages and regions parts correctly to find the most appropriate locale. WDYT?
See https://en.wikipedia.org/wiki/IETF_language_tag or https://andreipfeiffer.dev/slides/i18n-and-l10n.pdf

Something like as follows

```php
final class LanguageTag
{
    public string $language;
    public ?string $script = null;
    public ?string $region = null;
    public array $variants = [];
    public array $extensions = [];

    public function __construct(string $languageTag)
    {
        $components = explode('-', strtolower($languageTag));
        $this->language = array_shift($components);

        foreach ($components as $component) {
            if (strlen($component) === 4) {
                $this->script = ucfirst($component);
            } elseif (strlen($component) === 2) {
                $this->region = mb_strtoupper($component);
            }
            // Variants and extensions to handle
        }
    }
}
```

Reply:
I'm wondering the right place for this class, Translation or Intl? @Nyholm @OskarStark any thoughts?

Reply:
Note that for this PR we only need to capture the main parts. A regex such à `/^([a-z]{2,3})(?:-([a-z]{4}))?(?:-([a-z]{2}))?/` should be fine.

Reply:
Hi,

I completely reviewed the method and it now able to differentiate locales with language/script/region.
Also, I added some tests to make sure the selection is more accurate.

Reply:
> I'm wondering the right place for this class, Translation or Intl? @Nyholm @OskarStark any thoughts?

My feeling is about intl, but not sure

cc @stof

Reply:
I will start working on this soon. I'll create it in the Intl component and see if it needs to be moved elsewhere as it progresses.

Reply:
That is what intl provide (the PHP extension) with the [Locale](https://www.php.net/manual/en/class.locale.php) classes.

symfony/intl is kinda polyfill for those without intl extension. You may want to look at the int-contracts, but i really do not think symfony/intl should be the place for that.

As nowadays it contains massive amount of data... and finally not much class / helpers.

Reply:
OK thanks. I will reconsider this.
For this PR, there is no need to depend on other components.